src/main/java/com/arpnetworking/metrics/mad/parsers/ProtobufV2ToRecordParser.java

```diff
@@ -0,0 +1,138 @@
+        processEntries(metrics, record.getGaugesList(), MetricType.GAUGE);
+        return metrics.build();
+    }
+
+    private void processEntries(
+            final ImmutableMap.Builder<String, Metric> metrics,
+            final List<ClientV2.MetricEntry> entries,
+            final MetricType metricType) {
+        for (final ClientV2.MetricEntry metricEntry : entries) {
+            final DefaultMetric.Builder metricBuilder = new DefaultMetric.Builder()
+                    .setType(metricType);
+            final List<Quantity> quantities = Lists.newArrayListWithExpectedSize(metricEntry.getSamplesCount());
+            for (final ClientV2.Quantity quantity : metricEntry.getSamplesList()) {
+                quantities.add(
+                        new Quantity.Builder()
+                                .setUnit(baseUnit(quantity.getUnit()))
+                                .setValue(quantity.getDoubleValue())
+                                .build());
+            }
+
+            metricBuilder.setValues(quantities);
+            metrics.put(metricEntry.getName(), metricBuilder.build());
+        }
+    }
+
+    @Nullable
+    private Unit baseUnit(final ClientV2.CompoundUnit compoundUnit) {
+        if (!compoundUnit.getNumeratorList().isEmpty()) {
+            final ClientV2.Unit selectedUnit = compoundUnit.getNumerator(0);
+            if (ClientV2.Unit.Type.Value.UNRECOGNIZED.equals(selectedUnit.getType())
+                    || ClientV2.Unit.Type.Value.UNKNOWN.equals(selectedUnit.getType())) {
+                return null;
+            }
+            final String unitName;
+            if (!ClientV2.Unit.Scale.Value.UNIT.equals(selectedUnit.getScale())
+                    && !ClientV2.Unit.Scale.Value.UNRECOGNIZED.equals(selectedUnit.getScale())
+                    && !ClientV2.Unit.Scale.Value.UNKNOWN.equals(selectedUnit.getScale())) {
+                unitName = selectedUnit.getScale().name() + selectedUnit.getType().name();
+            } else {
+                unitName = selectedUnit.getType().name();
+            }
+
+            return Unit.valueOf(unitName);
+        }
+        return null;
+    }
+
+    private ImmutableMap<String, String> buildAnnotations(final ClientV2.Record record) {
+        final ImmutableMap.Builder<String, String> annotations = ImmutableMap.builder();
+        for (final ClientV2.AnnotationEntry annotationEntry : record.getAnnotationsList()) {
+            annotations.put(annotationEntry.getName(), annotationEntry.getValue());
+        }
+        return annotations.build();
+    }
+
+    private ImmutableMap<String, String> buildDimensions(final ClientV2.Record record) {
+        final ImmutableMap.Builder<String, String> dimensions = ImmutableMap.builder();
+        for (final ClientV2.DimensionEntry dimensionEntry : record.getDimensionsList()) {
+            dimensions.put(dimensionEntry.getName(), dimensionEntry.getValue());
+        }
+        return dimensions.build();
+    }
+}
```

Comment:
Do we not have any tests for v1? We need to add tests for v2.

Reply:
Added some tests for you.